File: Behavioral/State/Kotlin/Thread.kt
interface IThreadState
{
    fun Run(context: ThreadContext)
    fun Stop(context: ThreadContext)
    fun Wait(context: ThreadContext)
}

class ThreadContext
{
    var state: IThreadState

    init {
        println("Thread yaratildi ve henuz calismiyor")
        state = StoppedState()
    }

    fun Start() {
        state.Run(this)
    }

    fun Abort() {
        state.Stop(this)
    }

    fun Sleep() {
        state.Wait(this)
    }

}

class RunningState: IThreadState
{
    override fun Run(context: ThreadContext)
    {
        println("Thread zaten calisir durumda")
    }

    override fun Stop(context: ThreadContext)
    {
        context.state = StoppedState()
        println("Thread durduruldu")
    }

    override fun Wait(context: ThreadContext)
    {
        context.state = WaitingState()
        println("Thread gecici olarak cizelgeleme disina cikarildi")
    }
}

class StoppedState: IThreadState
{
    override fun Run(context: ThreadContext)
    {
        context.state = RunningState()
        println("Thread calismaya basladi")
    }

    override fun Stop(context: ThreadContext)
    {
        println("Thread zaten durdurulmus")
    }

    override fun Wait(context: ThreadContext)
    {
        println("Duran bir thread beklemeye alinamaz..")
    }
}

class WaitingState: IThreadState
{
    override fun Run(context: ThreadContext)
    {
        context.state = RunningState()
        println("Thread beklemedeyken tekrardan cizelgelemeye sokuldui")
    }

    override fun Stop(context: ThreadContext)
    {
        context.state = StoppedState();
        println("Beklmedeki thread tumuyle durduruldu")
    }

    override fun Wait(context: ThreadContext)
    {
        println("Thread zaten beklemede")
    }
}


// test
object Program
{
    @JvmStatic fun main(args:Array<String>)
    {
        val thread = ThreadContext()
        thread.Start()
        thread.Start()
        thread.Sleep()
        thread.Start()
        thread.Abort()
        thread.Start()
        thread.Abort()
        thread.Abort()
    }
}

File: Behavioral/State/Swift/Thread.swift
import Foundation

protocol IThreadState
{
    func Run(context:ThreadContext)
    func Stop(context:ThreadContext)
    func Wait(context:ThreadContext)
}

 class ThreadContext
 {
    private var state:IThreadState

    var State:IThreadState
    {
        get
        {
            return state
        }
        set(val)
        {
            state = val
        }
    }

    init() {
        print("Thread yaratildi ve henuz calismiyor")
        state = StoppedState()
    }

    func Start() {
        state.Run(context: self)
    }

    func Abort() {
        state.Stop(context: self)
    }

    func Sleep() {
        state.Wait(context: self)
    }

 }

 class RunningState: IThreadState
 {
    func Run(context:ThreadContext)
    {
        print("Thread zaten calisir durumda")
    }

    func Stop(context:ThreadContext)
    {
        context.State = StoppedState()
        print("Thread durduruldu")
    }

    func Wait(context:ThreadContext)
    {
        context.State = WaitingState()
        print("Thread gecici olarak cizelgeleme disina cikarildi")
    }
 }

class StoppedState: IThreadState
 {
    func Run(context:ThreadContext)
    {
        context.State = RunningState()
        print("Thread calismaya basladi")
    }

    func Stop(context:ThreadContext)
    {
        print("Thread zaten durdurulmus")
    }

    func Wait(context:ThreadContext)
    {
        print("Duran bir thread beklemeye alinamaz..")
    }
 }

 class WaitingState: IThreadState
 {
    func Run(context:ThreadContext)
    {
        context.State = RunningState()
        print("Thread beklemedeyken tekrardan cizelgelemeye sokuldui")
    }

    func Stop(context:ThreadContext)
    {
        context.State = StoppedState();
        print("Beklmedeki thread tumuyle durduruldu")
    }

    func Wait(context:ThreadContext)
    {
        print("Thread zaten beklemede")
    }
 }

 /////// test ///////
var thread = ThreadContext()
thread.Start()
thread.Start()
thread.Sleep()
thread.Start()
thread.Abort()
thread.Start()
thread.Abort()
thread.Abort()